JoinColony/colonyCDapp
# Feat: Send notifications for reputation motions

## Description

- Notifications should be sent for motions using reputation as a decision method.

These notifications should be sent for each lifecycle event of reputation motions:

```
- Created
    - Trigger: When a Motion is created and staked at least 10% of required stake (also send a mention for any recipients).
    - Event: `MotionStaked` (relevant only if 10% stake has been reached).
- Supported
    - Trigger: Motion has been fully staked.
    - Event: `MotionStaked` (needs to determine if fully staked).
- Opposed
    - Trigger: Motion has been fully opposed.
    - Event: `MotionStaked` (needs to determine if fully opposed).
- Voting
    - Trigger: Motion voting process has started.
    - Event: Motion state change.
- Reveal
    - Trigger: Motion reveal process has started.
    - Event: Motion state change.
- Finalize
    - Trigger: Motion has been finalized.
    - Event: `MotionFinalized`
```

Note: This doesn't cover motions made for funding an expenditure. You may also see in the specs a `Finalizable` notification for motions. That is also not included in this particular PR - only the ones above are.

There is also a block ingestor part of this PR found here: https://github.com/JoinColony/block-ingestor/pull/280

## Testing

Basically we want to test that the correct notifications are sent for motions, so the best way to test is to create some motions and go through the lifecycle of them. For example:

* Install the reputation voting extension.
<img width="1415" alt="Screenshot 2024-10-10 at 11 47 03" src="https://github.com/user-attachments/assets/54e2e815-7cd1-42d0-b595-ce4b9f13c64e">

* Make an action using reputation as the decision method, and stake it so that at least 10% support is achieved. This is the trigger for the action to become visible in the activity table, and also for notifications (colony-wide and mentions) to be sent for this motion. Check that these notifications come in (my example does not fire a mention, but anything targeting a user or users should)
<img width="1058" alt="Screenshot 2024-10-10 at 11 48 41" src="https://github.com/user-attachments/assets/9a6577d2-b0b2-4c28-8f2d-01eed10bce2b">
<img width="444" alt="Screenshot 2024-10-10 at 11 48 49" src="https://github.com/user-attachments/assets/0561cdfa-c661-4fe3-bac2-c4ee77bbb6f5">

* Support the motion to 100% by staking the max. Check that a `supported, will pass` notification arrives:
<img width="325" alt="Screenshot 2024-10-10 at 11 49 03" src="https://github.com/user-attachments/assets/bb98d745-a61e-4a93-9cb7-59a0a6903091">
<img width="451" alt="Screenshot 2024-10-10 at 11 49 17" src="https://github.com/user-attachments/assets/0959db69-fd7e-4551-b4df-31aec81f58e2">

* Now oppose the motion to 100% also, to trigger the voting. This should also fire a notification:
<img width="357" alt="Screenshot 2024-10-10 at 11 49 33" src="https://github.com/user-attachments/assets/f09783b0-d832-44f5-8f88-e136c5d2dd00">
<img width="444" alt="Screenshot 2024-10-10 at 11 49 39" src="https://github.com/user-attachments/assets/f9605f24-83b1-45c6-aac4-54e4e2e1b0f1">

* Vote to support the motion. After a couple of seconds, you should then be moved onto the reveal step, and a notification should be sent to show this:
<img width="363" alt="Screenshot 2024-10-10 at 11 50 07" src="https://github.com/user-attachments/assets/56ea9f74-fbdc-46b2-8bf2-a279ba249080">
<img width="445" alt="Screenshot 2024-10-10 at 11 50 12" src="https://github.com/user-attachments/assets/2843b2d5-f463-45dd-94b6-b8985e3f52f0">

* Finalize the motion, and check that a notification was sent:
<img width="443" alt="Screenshot 2024-10-10 at 11 51 48" src="https://github.com/user-attachments/assets/ed505801-aa44-4a43-9c97-7faf7fe6ff43">

* Try creating another motion, and this time immediately stake it to 100% opposed. The notifications should fire (one for the motion creation, and one for the opposition):
<img width="1049" alt="Screenshot 2024-10-10 at 11 53 52" src="https://github.com/user-attachments/assets/324dec50-6962-49fe-8905-70e45d04400c">
<img width="443" alt="Screenshot 2024-10-10 at 11 54 00" src="https://github.com/user-attachments/assets/2aabebc3-dec9-4760-90e2-71bf13ec46ce">

It would be great to try testing this with a couple of different types of actions / motions, if you have the time.

## Diffs

**New stuff** ✨

* New component: `MotionNotificationMessage`

**Changes** 🏗

* Moved NotificationType into graphql to be shared across the cdapp and block ingestor.
* Refactored the notification messages a little bit.
* Block ingestor changes: https://github.com/JoinColony/block-ingestor/pull/280

Resolves #3220
